server/s3_list_object_versions_action.cc

```diff
@@ -0,0 +1,835 @@
+    set_s3_error("ServiceUnavailable");
+  } else {
+    set_s3_error("InternalError");
+  }
+  send_response_to_s3_client();
+}
+
+void S3ListObjectVersionsAction::validate_request() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+
+  // Fetch input parameters
+  bucket_name = request->get_bucket_name();
+  s3_log(S3_LOG_DEBUG, request_id, "bucket name = %s\n", bucket_name.c_str());
+  request_prefix = request->get_query_string_value("prefix");
+  s3_log(S3_LOG_DEBUG, request_id, "prefix = %s\n", request_prefix.c_str());
+  request_delimiter = request->get_query_string_value("delimiter");
+  s3_log(S3_LOG_DEBUG, request_id, "delimiter = %s\n",
+         request_delimiter.c_str());
+  request_key_marker = request->get_query_string_value("key-marker");
+  s3_log(S3_LOG_DEBUG, request_id, "request_key_marker = %s\n",
+         request_key_marker.c_str());
+
+  last_key = request_key_marker;  // As requested by user
+  std::string max_k = request->get_query_string_value("max-keys");
+  if (max_k.empty()) {
+    max_keys = 1000;
+  } else {
+    if (!S3CommonUtilities::stoul(max_k, max_keys)) {
+      s3_log(S3_LOG_DEBUG, request_id, "invalid max-keys = %s\n",
+             max_k.c_str());
+      set_s3_error("InvalidArgument");
+      send_response_to_s3_client();
+      return;
+    }
+  }
+  s3_log(S3_LOG_DEBUG, request_id, "max-keys = %s\n", max_k.c_str());
+  next();
+}
+
+void S3ListObjectVersionsAction::get_next_versions() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  // If client is disconnected (say, due to read timeout, when
+  // the list object versions takes substantial time), destroy action class.
+  if (!request->client_connected()) {
+    s3_log(S3_LOG_INFO, stripped_request_id,
+           "s3 client is disconnected. Terminating list object versions "
+           "request\n");
+    S3_RESET_SHUTDOWN_SIGNAL;  // For shutdown testcases
+    done();
+    return;
+  }
+
+  const auto& object_version_list_index_layout =
+      bucket_metadata->get_objects_version_list_index_layout();
+
+  if (motr_kv_reader == nullptr) {
+    motr_kv_reader = s3_motr_kvs_reader_factory->create_motr_kvs_reader(
+        request, s3_motr_api);
+  }
+  // If keyval read operation failed due rpc message size threshold,
+  // try with lesser size.
+  if (motr_kv_reader->get_state() == S3MotrKVSReaderOpState::failed_e2big) {
+    if (retry_count > MAX_RETRY_COUNT) {
+      max_record_count = 1;
+    } else {
+      max_record_count = max_record_count / 2;
+    }
+  } else {
+    max_record_count = S3Option::get_instance()->get_motr_idx_fetch_count();
+  }
+  s3_log(S3_LOG_DEBUG, stripped_request_id, "max_record_count set to %zu\n",
+         max_record_count);
+
+  if (max_keys == 0) {
+    // As requested max_keys is 0
+    // Go ahead and respond.
+    fetch_successful = true;
+    send_response_to_s3_client();
+  } else if (zero(object_version_list_index_layout.oid)) {
```

Comment:
When does this happen?

Reply:
I guess when the version list table is not populated.